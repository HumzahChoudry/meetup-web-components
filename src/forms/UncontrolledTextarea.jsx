// @flow
import * as React from 'react';
import Textarea from './Textarea';

type State = {
	value: string,
};

type Props = {
	onChange?: (e: SyntheticInputEvent<*>) => null,
	value?: string,
};

/**
 * Should override value with info from state
 * @return {Object} the new state for the component
 */
export const overrideValue = (nextProps: Props): State => ({
	value: nextProps.value || '',
});

/**
 * Deprecated - use <Textarea /> directly, supply `value` from parent
 * @module UncontrolledTextarea
 * @deprecated
 */
class UncontrolledTextarea extends React.Component<Props, State> {
	state = {
		value: '',
	};

	/**
	 * @param {Object} nextProps the incoming props
	 * @return {undefined} side effect only
	 */
	static getDerivedStateFromProps(nextProps: Props) {
		return overrideValue(nextProps);
	}

	/**
	 * called as user changes value, updates state with new value
	 * @param  {Object} e Event object
	 * @return {undefined}
	 */
	onChange = (e: SyntheticInputEvent<EventTarget>): void => {
		const { onChange } = this.props;
		const { value } = e.target;

		this.setState(() => ({
			value,
		}));

		if (onChange) {
			onChange(e);
		}
	};

	render() {
		return (
			<Textarea {...this.props} value={this.state.value} onChange={this.onChange} />
		);
	}
}

UncontrolledTextarea.displayName = 'Textarea';

export default UncontrolledTextarea;
