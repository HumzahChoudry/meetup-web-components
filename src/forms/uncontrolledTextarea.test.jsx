import React from 'react';
import UncontrolledTextarea, { overrideValue } from './UncontrolledTextarea';
import { shallow } from 'enzyme';

describe('UncontrolledTextarea', () => {
	const props = {
		value: 'Batman',
		name: 'superhero',
		id: 'super',
		rows: 3,
		minHeight: 100,
		maxHeight: 300,
	};

	it('should set its value on input change', () => {
		const component = shallow(<UncontrolledTextarea {...props} required />);
		expect(component).toMatchSnapshot();
		component.setState({ value: 'New text value!' });
		expect(component).toMatchSnapshot();
	});

	it('should set state to value on `overrideValue`', () => {
		const newValue = 'someValue';
		const newState = overrideValue({ value: newValue });

		expect(newState.value).toEqual(newValue);
	});

	it('should set state to value on `overrideValue`', () => {
		const component = shallow(<UncontrolledTextarea {...props} required />);
		component.setProps({ value: undefined });
		expect(component.state().value).toBe('');
		expect(overrideValue({})).toEqual({ value: '' });
	});
});
