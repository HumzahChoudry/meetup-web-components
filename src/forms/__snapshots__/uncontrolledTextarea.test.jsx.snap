// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UncontrolledTextarea should set its value on input change 1`] = `
<WithErrorList(Textarea)
  id="super"
  maxHeight={300}
  minHeight={100}
  name="superhero"
  onChange={[Function]}
  required={true}
  rows={3}
  value="Batman"
/>
`;

exports[`UncontrolledTextarea should set its value on input change 2`] = `
<WithErrorList(Textarea)
  id="super"
  maxHeight={300}
  minHeight={100}
  name="superhero"
  onChange={[Function]}
  required={true}
  rows={3}
  value="Batman"
/>
`;
